libraries/bot-applicationinsights/src/test/java/com/microsoft/bot/applicationinsights/TelemetryWaterfallTests.java

```diff
@@ -0,0 +1,396 @@
+import com.microsoft.bot.builder.adapters.TestFlow;
+import com.microsoft.bot.dialogs.Dialog;
+import com.microsoft.bot.dialogs.DialogContext;
+import com.microsoft.bot.dialogs.DialogInstance;
+import com.microsoft.bot.dialogs.DialogReason;
+import com.microsoft.bot.dialogs.DialogSet;
+import com.microsoft.bot.dialogs.DialogState;
+import com.microsoft.bot.dialogs.WaterfallDialog;
+import com.microsoft.bot.dialogs.WaterfallStep;
+import org.junit.Assert;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mockito;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import static org.mockito.Mockito.*;
+
+public class TelemetryWaterfallTests {
+    @Captor
+    ArgumentCaptor<String> eventNameCaptor = ArgumentCaptor.forClass(String.class);
+
+    @Captor
+    ArgumentCaptor<Map<String, String>> propertiesCaptor = ArgumentCaptor.forClass(Map.class);
+
+    @Captor
+    ArgumentCaptor<Map<String, Double>> metricsCaptor = ArgumentCaptor.forClass(Map.class);
+
+    @Test
+    public void waterfall() {
+        ConversationState convoState = new ConversationState(new MemoryStorage());
+        TestAdapter adapter = new TestAdapter(TestAdapter.createConversationReference("Waterfall", "User1", "Bot"))
+            .use(new AutoSaveStateMiddleware(convoState));
+
+        BotTelemetryClient telemetryClient = Mockito.mock(BotTelemetryClient.class);
+        StatePropertyAccessor<DialogState> dialogState = convoState.createProperty("dialogState");
+        DialogSet dialogs = new DialogSet(dialogState);
+
+        dialogs.add(new WaterfallDialog("test", newWaterfall()));
+        dialogs.setTelemetryClient(telemetryClient);
+
+        new TestFlow(adapter, turnContext -> {
+            DialogContext dc = dialogs.createContext(turnContext).join();
+            dc.continueDialog().join();
+            if (!turnContext.getResponded()) {
+                dc.beginDialog("test", null).join();
+            }
+            return CompletableFuture.completedFuture(null);
+        })
+        .send("hello")
+        .assertReply("step1")
+        .send("hello")
+        .assertReply("step2")
+        .send("hello")
+        .assertReply("step3")
+        .startTest()
+        .join();
+
+        // C#'s trackEvent method of BotTelemetryClient has nullable parameters,
+        // therefore it always calls the same method.
+        // On the other hand, Java's BotTelemetryClient overloads the trackEvent method,
+        // so instead of calling the same method, it calls a method with less parameters.
+        // In this particular test, WaterfallDialog's beginDialog calls the method with only two parameters
+        verify(telemetryClient, times(4)).trackEvent(
+            anyString(),
+            anyMap()
+        );
```

Comment:
Mm, not quite sure about this, check another example in [Java](https://github.com/microsoft/botbuilder-java/blob/main/libraries/bot-ai-qna/src/test/java/com/microsoft/bot/ai/qna/QnAMakerTests.java#L1450) we did

Reply:
That's different because I am sure that test calls the overloaded with 3 arguments. This test calls the trackevent with 2 arguments, since its overloaded they are essentially two different methods.

Reply:
+1. I understood your point